### ISSUE TITLE
Investigate recent advances for next model backbone (2023/2024)

### ISSUE THREAD

**Issue by wingman-jr-addon** [wingman-jr-addon/model#7]
https://github.com/facebookresearch/ConvNeXt-V2/issues/3
https://github.com/edwardyehuang/iSeg/tree/master/backbones

**Comment by wingman-jr-addon** [comment]
Found an excellent resource for model implementations at https://github.com/leondgarse/keras_cv_attention_models#recognition-models, which should accelerate trying out new models. 

I've been doing some catch up on recent advancements of the past two or so years. While ConvNextV2 is intriguing due to model size, I think a better approach for this use case will be to not only focus on the model size but also on the pretraining. In particular, I think the DINO/DINOv2 and CLIP-related pretraining approaches are particularly helpful due to the robustness in distribution shift. Not only is the training material much closer to our target distribution, but generally the models will be much stronger.

To test this theory out, I tried a DINOv2 finetune out (dense layer only + gradual weight change on last 15 layers) and got excellent results, better than I had seen for some of my more half-hearted approaches to e.g. Inceptions and/or ResNets. The only challenge there is that the smallest model available uses a whopping 47.23G FLOPS (vs. 0.72G for say EfficientNetV1 B0). A bit surprisingly, I was able to successfully convert it to TensorflowJS but it was slow to the level of several seconds per image prediction. Still, a useful experiment to demonstrate effectiveness of a stronger model. The dataset has also increased somewhat, so it's not quite apples to apples but notice the improvement in the DET and ROC curves.
SQRXR 112 (EfficientNet Lite L0-based):
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/f7636a3b-c2fe-4d51-a9d2-71e93598c22d)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/46792a8a-6b61-4ed7-bf1d-5eee6c1ff23e)

SQRXR 119 (DINOv2):
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/ceac4cfc-7de4-42e8-af40-6d0657a27722)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/9b208e43-a10d-4cf9-a9c9-60a89e69d7c9)

I'm going to check out an EVA02 based model next as it is CLIP-based, but it still weighs in at 4.72G FLOPS so it's in theory going to be a few times slower than the current.

**Comment by wingman-jr-addon** [comment]
So I tried out the smallest EVA02, EVA02TinyPatch14. I tried training with various finetunes changing the number of layers of the graph I retuned. Results were OK, but final DET graphs showed poor and/or uneven performance. My hypothesis is to try the next size up EVA02 and see if I gain more of the smoothness and performance I observed in the megasized DINOv2.

For comparison, here's a DET output from EVA02Tiny (SQRXR 120):
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/13ff9bc2-5e52-4589-b0d7-88e5cc37e77b)

Now compare that to the current deployed EfficientNetLite L0 based approach (SQRXR 119):
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/beb7376f-7483-4a03-b4e1-1b2004a87ba1)

**Comment by wingman-jr-addon** [comment]
Results from EVA02 larger model were decent but not enough to justify performance penalty.
SQRXR 121 EVA02Small:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/76723f86-c8e0-4121-a5fa-59f63868349d)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/26536e87-ca02-4c4b-9fcc-9e98c2a9568b)

**Comment by wingman-jr-addon** [comment]
EfficientFormerV2S2 seems like a potential for incremental improvement. (SQRXR 122)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/2ceb4a32-9f19-407d-900b-cdb192f3c815)
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/402653de-7e5a-445e-b4a9-ee855d19f9ed)

Still a little squiggly on the bottom of the DET curve, not such a fan for FPR on the "trusted" zone. Still good though and inference only increased from about 68ms for SQRXR 112 to 82ms for this one SQRXR 122.

**Comment by wingman-jr-addon** [comment]
I had a few more experiments.

1. I tried playing around with Uniformer. However, I was getting reasonable but much lower than expected accuracy so I suspect I'm not integrating something correctly.
2. I worked with EdgeNext_Base. DET wasn't as promising as it could be:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/6d41a02d-7278-42c0-aa29-08cb47fa9cda)
3. I returned to an EfficientFormer variant, EfficientFormerV2L. Performance was not better than current, and was actually no better than EfficientFormerV2S2, which surprised me:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/3d352d1a-de75-4140-a2d7-3715c9394c7f)
4. I returned to the possible incremental results from EfficientFormerV2S2 and looked at ways to smooth out the DET curve by playing with training, making some updates like a switch to AdamW in a couple places. This was successful. SQRXR 127 (compare to SQRXR 122 in last comment):
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/adee4d03-97e6-45aa-b0c7-63959521f64d)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/a4656ec4-dd70-444d-8258-fe283d3213f9)
It's a subtle improvement over SQRXR 112 because the DET curve bows in slightly. For example, 5% FNR crosses 20% FPR on SQRXR 112 but SQRXR 127 is clearly below. The non-linear scaling is something to watch carefully here as subtle changes can actually mean bigger changes in the final model performance.

**Comment by wingman-jr-addon** [comment]
I tried working with EfficientViT B2 as SQRXR 128. Training went well and overall results were promising:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/e574b158-92bf-4b9f-b167-8d0df1217865)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/ce5bdb23-4408-4f05-9d3a-34ffb40686c0)
Unfortunately the resulting model was a bit difficult to both reload as well as to convert into TF.js. The use of 'hard_swish' did not play well, but I was able to coax the data into a custom layer instead of a function and got it to reload; however, the use of the PartitionedCall op ultimately meant TF.js couldn't handle it. Might be something to return to as there may be a way to coax the model to not have a PartitionedCall at some point but it's not obvious.

Next up trying RepViT M11 as SQRXR 129. Training was OK but did not seem to provide much advantage over say SQRXR 127.
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/6d09d756-551b-4682-8e6c-be30669b2465)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/1d4183f3-1308-4aec-9aea-190ae9c49a81)

**Comment by wingman-jr-addon** [comment]
Next - Levit 256 as SQRXR 130:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/93cf3d55-aa7e-4046-bf6e-0b7fcbe3bab2)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/a1103ef2-2b44-4c7d-802a-7b72f95649d5)
Marginal advantage on ROC over baseline 112, less good DET.

**Comment by wingman-jr-addon** [comment]
CMT XS Torch as SQRXR 131:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/73142bc0-9ceb-4d49-9971-6b1df3b63b00)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/b87640c6-8244-40f2-8efc-7b98057efdc3)
Marginal advantage on ROC AUC, disadvantage on DET.

**Comment by wingman-jr-addon** [comment]
TinyViT 11 as SQRXR 132:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/661390d1-7627-45d3-aaf0-a0e5b01c91ca)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/9c61042f-12b5-44f9-ae59-c6f2815ad5b0)
Marginal advantage on ROC AUC, disadvantage on DET.

**Comment by wingman-jr-addon** [comment]
EfficientNetV2 B0 as SQRXR 133:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/735bccdc-16e9-41af-a71b-3ef1becf7a92)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/21d0feb7-a606-4c3d-b561-ae0ae777e799)
No advantage.

**Comment by wingman-jr-addon** [comment]
EfficientFormerV2S0 as a smaller variant of an earlier experiment. SQRXR 134:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/05421b87-273f-4b75-b222-05e6c2cfae59)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/53412482-d935-4781-baa0-c729b81b10c8)
No improvement, and unsurprisingly not as good as V2S2.

**Comment by wingman-jr-addon** [comment]
Tried a bit different training regime with current EfficientNetLite L0 using some of the other advances like swapping out for AdamW. SQRXR 135:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/bcdc1476-e47f-4116-9ae8-8525fe6de3ac)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/3fa1c4ce-e692-4905-ab67-9feafaccdd64)
About the same but the DET curve's a bit more gnarly at the beginning. No clear advantage. Still, I think there may be something to the training technique approach here.

**Comment by wingman-jr-addon** [comment]
GCViT XTiny - a bit bigger model. Shows in the performance. SQRXR 136:
![post_det](https://github.com/wingman-jr-addon/model/assets/55339824/cf942e99-8abe-44d4-a6bf-ca215a5721aa)
![post_roc](https://github.com/wingman-jr-addon/model/assets/55339824/ba278e81-e35f-477f-babf-38e0a1ddc7fe)
Definite improvement. Scanning speed is slow but sort-of-tolerable. Might be useful as a bigger model. The adventurous can try it out on the test branch while it sticks around: https://github.com/wingman-jr-addon/wingman_jr/tree/sqrxr-136

**Comment by wingman-jr-addon** [comment]
I've been trying this out ... and I'm not sure it's fast enough or good enough to become the next top model yet. I might need to keep searching.

**Comment by Ro-Den** [comment]
> CLIP-related pretraining approaches are particularly helpful due to the robustness in distribution shift. Not only is the training material much closer to our target distribution, but generally the models will be much stronger.

What's the point of using Contrastive Language-Image Pretraining (CLIP) for simple (multi-class) classification tasks?

It's main usage is [[de]generative AI aka diffusion );](https://www.youtube.com/watch?v=_7rMfsA24Ls&t=6426s)

I'd rather investigate image regression or object detection...

**Comment by wingman-jr-addon** [comment]
@Ro-Den well, the sentence you highlighted is actually precisely the reason. While most of the time CLIP is viewed - as you noted - for tasks combining imagery and text, a side effect of the essentially semi-supervised training that CLIP does (yes I know that's a bit inaccurate), is that it tends to be much more robust in distribution shift. In a practical sense that means that while a CLIP-based model and another model both might do well on the distribution of images I have in my training and test sets, that we would expect a CLIP-based model to perform better in the real-world - even given the same outcomes on the test set. That's quite appealing.

(EDIT: As an example, see https://arxiv.org/pdf/2410.01534v1 , with various findings like 
page 4-5:

> (1) CLIP models are generally more robust than other
> ImageNet models on six out of ten visual factors. Fig. 1
> highlights several insights into the factor-level robustness of
> CLIP models. First, we find that CLIP models are more robust
> than other models on six out of ten visual factors, including
> Subcategory, Smaller, Color, Shape, Texture, and Larger.
> Specifically, CLIP models exhibit higher factor-level effective
> robustness than other models on each of these factors. Second,
> we observe that CLIP models are less robust than other models
> on Pose and Partial View. Third, CLIP models show a similar
> trend to other models on the Background factor. Moreover,
> Idrissi et al. [14] observe that data augmentations can improve
> robustness to related factors, but with spill-over effects to
> unrelated factors. We speculate that data augmentations used
> during CLIP training may cause similar effects.

page 9:

> (2) ID calibration of CLIP models transfers to OOD
> test sets. While prior studies [90] report in-distribution (ID)
> calibration often fails to generalize under distribution shifts,
> our findings reveal a promising result for CLIP models. After
> calibrating CLIP models on the ID set, they exhibit improved
> calibration on OOD test sets. For example, on ImageNet-A,
> CLIP models exhibit lower calibration error after temperature
> scaling, a trend not seen in other models. This suggests that
> CLIP models are relatively easier to calibrate across diverse
> distributions, indicating their potential for robust and reliable
> applications in real-world settings.
> 
)

Yes, I certainly have thought about object detection. But from my research, those models tend to be a bit heavier than just image classification - and I'm not planning to do anything extra with the identified area in the addon that this model drives. (Keep in mind that user feedback suggests speed of image scanning is perhaps the number 2 or 3 complaint) Maybe there's a win with better robustness or something though?

I had looked into image regression in the past, but at least say five years ago there was a sense that trying to map e.g. linear regressions out of images was that it didn't train the model nearly as well. That being said, I do essentially do a regression but the way I approach it is by 1) pre-training the core model using classification then 2) extract the features and training against a binary classifier optimizing for AUC ROC. See the description here: https://github.com/wingman-jr-addon/model/tree/master/sqrxr_112#the-sqrxr-second-stage-model

**Comment by wingman-jr-addon** [comment]
Closing this and moving to the 2025 issue here https://github.com/wingman-jr-addon/model/issues/9
